### DISCUSSION TITLE
Does a fresh app require `import_map.json` to be at the root of its directory?

### DISCUSSION THREAD

**Discussion by mmazzarolo** [freshframework/fresh#360] [Q&A] [answered]
Hola!
This might sound like a weird question, but: does a fresh app require `import_map.json` to be at the root of its directory? 
Can we we move the fresh app dir into another deno project and move/adjust its `import_map.json` accordingly?  

I'm working in a medium-size deno project that has a monorepo-like structure like this:

```bash
.
├── modules/
│   ├── server/
│   │   ├── a.ts
│   │   ├── b.ts
│   │   └── mod.ts
│   ├── library-one/
│   │   └── mod.ts
│   └── library-two/
│       └── mod.ts
├── deno.jsonc
└── import_map.json
```

In the root-level `import_map.json` I take care of mapping both local imports (`"library-one": "./modules/library-one/mod.ts"`) and remote ones (`"fs": "https://deno.land/std@0.145.0/fs/mod.ts"`).  
This approach has been working pretty well for me so far, and I wanted to try integrating a new Fresh app in the same flow. 

To do so, I:
- Created a new fresh app in `/modules/fresh-app`.
- Copy/pasted the content of `/modules/fresh-app/import_map.json` to my root-level `import_map.json`, and deleted Fresh's `import_map.json`. 
- Updated my root level `deno.jsonc` to kickstart the app (`"start-fresh-app": "deno run --import-map import_map.json -A --watch=./modules/fresh-app/static/,./modules/fresh-app/routes/ modules/fresh-app/dev.ts"`) and deleted Fresh's `deno.json`. 

I assumed this would work, but running `start-fresh-app` throws an error about a failed call to `deno info`:
```
error: Unable to load 'file:///Users/blabla/my-app/modules/fresh-app/import_map.json' import map

Caused by:
    No such file or directory (os error 2)
error: Unable to load 'file:///Users/blabla/my-app/modules/fresh-app/import_map.json' import map

error: Failed to call 'deno info' on 'https://deno.land/x/fresh@1.0.0/src/runtime/main.ts'
    at failureErrorWithLog (https://deno.land/x/esbuild@v0.14.39/mod.js:1572:15)
    at https://deno.land/x/esbuild@v0.14.39/mod.js:1218:28
    at runOnEndCallbacks (https://deno.land/x/esbuild@v0.14.39/mod.js:1001:63)
    at buildResponseToResult (https://deno.land/x/esbuild@v0.14.39/mod.js:1216:7)
    at https://deno.land/x/esbuild@v0.14.39/mod.js:1325:14
    at https://deno.land/x/esbuild@v0.14.39/mod.js:633:9
    at handleIncomingPacket (https://deno.land/x/esbuild@v0.14.39/mod.js:730:9)
    at readFromStdout (https://deno.land/x/esbuild@v0.14.39/mod.js:599:7)
    at https://deno.land/x/esbuild@v0.14.39/mod.js:1846:11
```

Restoring the original `/modules/fresh-app/import_map.json` is enough to fix the issue — it seems some dependency requires it to be at the root of the fresh-app?

It's not that big of deal for me, but before starting to investigate, **I wanted to ask here if there's any reason `import_map.json` needs to be at the root of the fresh app dir?** (I might also just be missing something on my end 😅).  

Oh, and btw, great work with Fresh (and Deno!)

**Comment by m4rocks** [comment]
not really an experienced deno user so take this with a grain of salt. from the code i read and what i played around with, i noticed that fresh itself grabs the `import_map.json` from the root of the project.

here is the line where I noticed this.
https://github.dev/denoland/fresh/blob/74387ff92bf8f8bfdbd9a0361e8f92816dc9e600/src/server/context.ts#L97

**Reply by mmazzarolo** [reply]
Hey! yeah, I'm pretty sure that's why it's happening.  
I would assume this path could be made configurable, but I'd understand that doing so is not a priority :)

**Comment by marvinhagemeister** [comment, marked as answer]
For anyone coming across this discussion: Since Deno `1.30.0` the `deno.json` file supports an `imports` field which is basically an inline import map. See the documentation for that here https://deno.land/manual@v1.36.0/getting_started/configuration_file#imports-and-scopes
